// @polkadot/types/primitive/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "BitVec", {
  enumerable: true,
  get: function () {
    return _BitVec.BitVec;
  }
});
Object.defineProperty(exports, "bool", {
  enumerable: true,
  get: function () {
    return _Bool.bool;
  }
});
Object.defineProperty(exports, "Bool", {
  enumerable: true,
  get: function () {
    return _Bool.bool;
  }
});
Object.defineProperty(exports, "Bytes", {
  enumerable: true,
  get: function () {
    return _Bytes.Bytes;
  }
});
Object.defineProperty(exports, "Data", {
  enumerable: true,
  get: function () {
    return _Data.Data;
  }
});
Object.defineProperty(exports, "DoNotConstruct", {
  enumerable: true,
  get: function () {
    return _DoNotConstruct.DoNotConstruct;
  }
});
Object.defineProperty(exports, "i8", {
  enumerable: true,
  get: function () {
    return _I.i8;
  }
});
Object.defineProperty(exports, "I8", {
  enumerable: true,
  get: function () {
    return _I.i8;
  }
});
Object.defineProperty(exports, "i16", {
  enumerable: true,
  get: function () {
    return _I2.i16;
  }
});
Object.defineProperty(exports, "I16", {
  enumerable: true,
  get: function () {
    return _I2.i16;
  }
});
Object.defineProperty(exports, "i32", {
  enumerable: true,
  get: function () {
    return _I3.i32;
  }
});
Object.defineProperty(exports, "I32", {
  enumerable: true,
  get: function () {
    return _I3.i32;
  }
});
Object.defineProperty(exports, "i64", {
  enumerable: true,
  get: function () {
    return _I4.i64;
  }
});
Object.defineProperty(exports, "I64", {
  enumerable: true,
  get: function () {
    return _I4.i64;
  }
});
Object.defineProperty(exports, "i128", {
  enumerable: true,
  get: function () {
    return _I5.i128;
  }
});
Object.defineProperty(exports, "I128", {
  enumerable: true,
  get: function () {
    return _I5.i128;
  }
});
Object.defineProperty(exports, "i256", {
  enumerable: true,
  get: function () {
    return _I6.i256;
  }
});
Object.defineProperty(exports, "I256", {
  enumerable: true,
  get: function () {
    return _I6.i256;
  }
});
Object.defineProperty(exports, "Null", {
  enumerable: true,
  get: function () {
    return _Null.Null;
  }
});
Object.defineProperty(exports, "StorageKey", {
  enumerable: true,
  get: function () {
    return _StorageKey.StorageKey;
  }
});
Object.defineProperty(exports, "Text", {
  enumerable: true,
  get: function () {
    return _Text.Text;
  }
});
Object.defineProperty(exports, "Type", {
  enumerable: true,
  get: function () {
    return _Type.Type;
  }
});
Object.defineProperty(exports, "u8", {
  enumerable: true,
  get: function () {
    return _U.u8;
  }
});
Object.defineProperty(exports, "U8", {
  enumerable: true,
  get: function () {
    return _U.u8;
  }
});
Object.defineProperty(exports, "u16", {
  enumerable: true,
  get: function () {
    return _U2.u16;
  }
});
Object.defineProperty(exports, "U16", {
  enumerable: true,
  get: function () {
    return _U2.u16;
  }
});
Object.defineProperty(exports, "u32", {
  enumerable: true,
  get: function () {
    return _U3.u32;
  }
});
Object.defineProperty(exports, "U32", {
  enumerable: true,
  get: function () {
    return _U3.u32;
  }
});
Object.defineProperty(exports, "u64", {
  enumerable: true,
  get: function () {
    return _U4.u64;
  }
});
Object.defineProperty(exports, "U64", {
  enumerable: true,
  get: function () {
    return _U4.u64;
  }
});
Object.defineProperty(exports, "u128", {
  enumerable: true,
  get: function () {
    return _U5.u128;
  }
});
Object.defineProperty(exports, "U128", {
  enumerable: true,
  get: function () {
    return _U5.u128;
  }
});
Object.defineProperty(exports, "u256", {
  enumerable: true,
  get: function () {
    return _U6.u256;
  }
});
Object.defineProperty(exports, "U256", {
  enumerable: true,
  get: function () {
    return _U6.u256;
  }
});
Object.defineProperty(exports, "usize", {
  enumerable: true,
  get: function () {
    return _USize.usize;
  }
});
Object.defineProperty(exports, "USize", {
  enumerable: true,
  get: function () {
    return _USize.usize;
  }
});

var _BitVec = require("./BitVec");

var _Bool = require("./Bool");

var _Bytes = require("./Bytes");

var _Data = require("./Data");

var _DoNotConstruct = require("./DoNotConstruct");

var _I = require("./I8");

var _I2 = require("./I16");

var _I3 = require("./I32");

var _I4 = require("./I64");

var _I5 = require("./I128");

var _I6 = require("./I256");

var _Null = require("./Null");

var _StorageKey = require("./StorageKey");

var _Text = require("./Text");

var _Type = require("./Type");

var _U = require("./U8");

var _U2 = require("./U16");

var _U3 = require("./U32");

var _U4 = require("./U64");

var _U5 = require("./U128");

var _U6 = require("./U256");

var _USize = require("./USize");
