// @polkadot/metadata/decorate/index.mjs
// Copyright 2017-2021 @polkadot/metadata authors & contributors
// SPDX-License-Identifier: Apache-2.0
import { assert } from '@polkadot/util';
import { Metadata } from "../Metadata.mjs";
import { decorateConstants } from "./constants/index.mjs";
import { decorateErrors } from "./errors/index.mjs";
import { decorateEvents } from "./events/index.mjs";
import { decorateExtrinsics } from "./extrinsics/index.mjs";
import { decorateStorage } from "./storage/index.mjs";
/**
 * Expands the metadata by decoration into consts, query and tx sections
 */

export function expandMetadata(registry, metadata) {
  assert(metadata instanceof Metadata, 'You need to pass a valid Metadata instance to Decorated');
  const latest = metadata.asLatest;
  return {
    consts: decorateConstants(registry, latest),
    errors: decorateErrors(registry, latest, metadata.version),
    events: decorateEvents(registry, latest, metadata.version),
    query: decorateStorage(registry, latest, metadata.version),
    tx: decorateExtrinsics(registry, latest, metadata.version)
  };
}
export { decorateConstants, decorateExtrinsics, decorateStorage };
