// @polkadot/types/create/encodeTypes.mjs
import _defineProperty from "@babel/runtime/helpers/esm/defineProperty";

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

// Copyright 2017-2021 @polkadot/types authors & contributors
// SPDX-License-Identifier: Apache-2.0
import { assert, isNumber, isUndefined } from '@polkadot/util';
import { TypeDefInfo } from "./types.mjs";

const stringIdentity = value => value.toString();

const INFO_WRAP = ['BTreeMap', 'BTreeSet', 'Compact', 'HashMap', 'Option', 'Result', 'Vec'];
export function paramsNotation(outer, inner, transform = stringIdentity) {
  return `${outer}${inner ? `<${(Array.isArray(inner) ? inner : [inner]).map(transform).join(', ')}>` : ''}`;
}

function encodeWithParams(typeDef, outer) {
  const {
    info,
    sub
  } = typeDef;

  switch (info) {
    case TypeDefInfo.BTreeMap:
    case TypeDefInfo.BTreeSet:
    case TypeDefInfo.Compact:
    case TypeDefInfo.HashMap:
    case TypeDefInfo.Linkage:
    case TypeDefInfo.Option:
    case TypeDefInfo.Result:
    case TypeDefInfo.Vec:
      return paramsNotation(outer, sub, param => encodeTypeDef(param));
  }

  throw new Error(`Unable to encode ${JSON.stringify(typeDef)} with params`);
}

function encodeDoNotConstruct({
  displayName
}) {
  return `DoNotConstruct<${displayName || 'Unknown'}>`;
}

function encodeSubTypes(sub, asEnum) {
  const names = sub.map(({
    name
  }) => name);
  assert(names.every(n => !!n), `Subtypes does not have consistent names, ${names.join(', ')}`);
  const inner = sub.reduce((result, type) => _objectSpread(_objectSpread({}, result), {}, {
    [type.name]: encodeTypeDef(type)
  }), {});
  return JSON.stringify(asEnum ? {
    _enum: inner
  } : inner);
}

function encodeEnum(typeDef) {
  assert(typeDef.sub && Array.isArray(typeDef.sub), 'Unable to encode Enum type');
  const sub = typeDef.sub; // c-like enums have all Null entries
  // TODO We need to take the disciminant into account and auto-add empty entries

  return sub.every(({
    type
  }) => type === 'Null') ? JSON.stringify({
    _enum: sub.map(({
      name
    }, index) => `${name || `Empty${index}`}`)
  }) : encodeSubTypes(sub, true);
}

function encodeStruct(typeDef) {
  assert(typeDef.sub && Array.isArray(typeDef.sub), 'Unable to encode Struct type');
  return encodeSubTypes(typeDef.sub);
}

function encodeTuple(typeDef) {
  assert(typeDef.sub && Array.isArray(typeDef.sub), 'Unable to encode Tuple type');
  return `(${typeDef.sub.map(type => encodeTypeDef(type)).join(', ')})`;
}

function encodeUInt({
  length
}, type) {
  assert(isNumber(length), 'Unable to encode VecFixed type');
  return `${type}<${length}>`;
}

function encodeVecFixed({
  length,
  sub
}) {
  assert(isNumber(length) && !isUndefined(sub) && !Array.isArray(sub), 'Unable to encode VecFixed type');
  return `[${sub.type};${length}]`;
} // We setup a record here to ensure we have comprehensive coverage (any item not covered will result
// in a compile-time error with the missing index)


const encoders = {
  [TypeDefInfo.BTreeMap]: typeDef => encodeWithParams(typeDef, 'BTreeMap'),
  [TypeDefInfo.BTreeSet]: typeDef => encodeWithParams(typeDef, 'BTreeSet'),
  [TypeDefInfo.Compact]: typeDef => encodeWithParams(typeDef, 'Compact'),
  [TypeDefInfo.DoNotConstruct]: typeDef => encodeDoNotConstruct(typeDef),
  [TypeDefInfo.Enum]: typeDef => encodeEnum(typeDef),
  [TypeDefInfo.HashMap]: typeDef => encodeWithParams(typeDef, 'HashMap'),
  [TypeDefInfo.Int]: typeDef => encodeUInt(typeDef, 'Int'),
  [TypeDefInfo.Linkage]: typeDef => encodeWithParams(typeDef, 'Linkage'),
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  [TypeDefInfo.Null]: typeDef => 'Null',
  [TypeDefInfo.Option]: typeDef => encodeWithParams(typeDef, 'Option'),
  [TypeDefInfo.Plain]: typeDef => typeDef.displayName || typeDef.type,
  [TypeDefInfo.Result]: typeDef => encodeWithParams(typeDef, 'Result'),
  [TypeDefInfo.Set]: typeDef => typeDef.type,
  [TypeDefInfo.Struct]: typeDef => encodeStruct(typeDef),
  [TypeDefInfo.Tuple]: typeDef => encodeTuple(typeDef),
  [TypeDefInfo.UInt]: typeDef => encodeUInt(typeDef, 'UInt'),
  [TypeDefInfo.Vec]: typeDef => encodeWithParams(typeDef, 'Vec'),
  [TypeDefInfo.VecFixed]: typeDef => encodeVecFixed(typeDef)
};

function encodeType(typeDef) {
  const encoder = encoders[typeDef.info];
  assert(encoder, `Cannot encode type: ${JSON.stringify(typeDef)}`);
  return encoder(typeDef);
}

export function encodeTypeDef(typeDef) {
  assert(!isUndefined(typeDef.info), `Invalid type definition with no instance info, ${JSON.stringify(typeDef)}`); // In the case of contracts we do have the unfortunate situation where the displayName would
  // refer to "Option" when it is an option. For these, string it out, only using when actually
  // not a top-level element to be used

  if (typeDef.displayName && !INFO_WRAP.some(i => typeDef.displayName === i)) {
    return typeDef.displayName;
  }

  return encodeType(typeDef);
}
export function withTypeString(typeDef) {
  return _objectSpread(_objectSpread({}, typeDef), {}, {
    type: encodeType(typeDef)
  });
}
