// @polkadot/types/generic/ConsensusEngineId.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.GenericConsensusEngineId = exports.CID_POW = exports.CID_GRPA = exports.CID_BABE = exports.CID_AURA = void 0;

var _bn = _interopRequireDefault(require("bn.js"));

var _util = require("@polkadot/util");

var _U = require("../primitive/U32");

// Copyright 2017-2021 @polkadot/types authors & contributors
// SPDX-License-Identifier: Apache-2.0
// there are all reversed since it is actually encoded as u32, LE,
// this means that FRNK has the bytes as KNRF
const CID_AURA = 0x61727561; // 'aura'

exports.CID_AURA = CID_AURA;
const CID_BABE = 0x45424142; // 'BABE'

exports.CID_BABE = CID_BABE;
const CID_GRPA = 0x4b4e5246; // 'FRNK' (don't ask, used to be afg1)

exports.CID_GRPA = CID_GRPA;
const CID_POW = 0x5f776f70; // 'pow_'

exports.CID_POW = CID_POW;

/**
 * @name GenericConsensusEngineId
 * @description
 * A 4-byte identifier (actually a [u8; 4]) identifying the engine, e.g. for Aura it would be [b'a', b'u', b'r', b'a']
 */
class GenericConsensusEngineId extends _U.u32 {
  static idToString(input) {
    return (0, _util.bnToBn)(input).toArray('le').map(code => String.fromCharCode(code)).join('');
  }

  static stringToId(input) {
    return input.split('').reverse().reduce((result, char) => result * 256 + char.charCodeAt(0), 0);
  }
  /**
   * @description `true` if the engine matches aura
   */


  get isAura() {
    return this.eq(CID_AURA);
  }
  /**
   * @description `true` is the engine matches babe
   */


  get isBabe() {
    return this.eq(CID_BABE);
  }
  /**
   * @description `true` is the engine matches grandpa
   */


  get isGrandpa() {
    return this.eq(CID_GRPA);
  }
  /**
   * @description `true` is the engine matches pow
   */


  get isPow() {
    return this.eq(CID_POW);
  }

  _getAuraAuthor(bytes, sessionValidators) {
    return sessionValidators[this.registry.createType('RawAuraPreDigest', bytes.toU8a(true)).slotNumber.mod(new _bn.default(sessionValidators.length)).toNumber()];
  }

  _getBabeAuthor(bytes, sessionValidators) {
    const digest = this.registry.createType('RawBabePreDigestCompat', bytes.toU8a(true));
    return sessionValidators[digest.value.toNumber()];
  }

  _getPowAuthor(bytes) {
    return this.registry.createType('AccountId', bytes);
  }
  /**
   * @description From the input bytes, decode into an author
   */


  extractAuthor(bytes, sessionValidators) {
    if (sessionValidators !== null && sessionValidators !== void 0 && sessionValidators.length) {
      if (this.isAura) {
        return this._getAuraAuthor(bytes, sessionValidators);
      } else if (this.isBabe) {
        return this._getBabeAuthor(bytes, sessionValidators);
      }
    }

    if (this.isPow) {
      return this._getPowAuthor(bytes);
    }

    return undefined;
  }
  /**
   * @description Override the default toString to return a 4-byte string
   */


  toString() {
    return GenericConsensusEngineId.idToString(this);
  }

}

exports.GenericConsensusEngineId = GenericConsensusEngineId;
