// @polkadot/types/generic/MultiAddress.mjs
// Copyright 2017-2021 @polkadot/types authors & contributors
// SPDX-License-Identifier: Apache-2.0
import { isBn, isNumber, isString, isU8a } from '@polkadot/util';
import { decodeAddress } from '@polkadot/util-crypto';
import { Enum } from "../codec/Enum.mjs";
import { GenericAccountId } from "./AccountId.mjs";
import { GenericAccountIndex } from "./AccountIndex.mjs";

function decodeMultiU8a(registry, value) {
  if (isU8a(value) && value.length <= 32) {
    if (value.length === 32) {
      return {
        id: value
      };
    } else if (value.length === 20) {
      return {
        Address20: value
      };
    } else {
      return decodeMultiAny(registry, registry.createType('AccountIndex', value));
    }
  }

  return value;
}

function decodeMultiAny(registry, value) {
  if (value instanceof GenericMultiAddress) {
    return value;
  } else if (value instanceof GenericAccountId) {
    return {
      Id: value
    };
  } else if (value instanceof GenericAccountIndex || isNumber(value) || isBn(value)) {
    return {
      Index: registry.createType('Compact<AccountIndex>', value)
    };
  } else if (isString(value)) {
    return decodeMultiU8a(registry, decodeAddress(value.toString()));
  }

  return decodeMultiU8a(registry, value);
}

export class GenericMultiAddress extends Enum {
  constructor(registry, value) {
    super(registry, {
      Id: 'AccountId',
      Index: 'Compact<AccountIndex>',
      Raw: 'Bytes',
      // eslint-disable-next-line sort-keys
      Address32: 'H256',
      // eslint-disable-next-line sort-keys
      Address20: 'H160'
    }, decodeMultiAny(registry, value));
  }
  /**
   * @description Returns the string representation of the value
   */


  toString() {
    return this.value.toString();
  }

}
