// @polkadot/types-known/upgrades/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _bn = _interopRequireDefault(require("bn.js"));

var _util = require("@polkadot/util");

var _kusama = _interopRequireDefault(require("./kusama"));

var _polkadot = _interopRequireDefault(require("./polkadot"));

var _westend = _interopRequireDefault(require("./westend"));

// Copyright 2017-2021 @polkadot/types-known authors & contributors
// SPDX-License-Identifier: Apache-2.0

/** @internal */
function checkOrder(network, versions) {
  const ooo = versions.filter((curr, index) => {
    const prev = versions[index - 1];
    return index === 0 ? false : curr[0] <= prev[0] || curr[1] <= prev[1];
  });
  (0, _util.assert)(!ooo.length, `${network}: Mismatched upgrade ordering: ${JSON.stringify(ooo)}`);
  return versions;
}
/** @internal */


function rawToFinal(network, {
  genesisHash,
  versions
}) {
  return {
    genesisHash: (0, _util.hexToU8a)(genesisHash),
    network,
    versions: checkOrder(network, versions).map(([blockNumber, specVersion]) => ({
      blockNumber: new _bn.default(blockNumber),
      specVersion: new _bn.default(specVersion)
    }))
  };
} // Type overrides for specific spec types & versions as given in runtimeVersion


const upgrades = [rawToFinal('kusama', _kusama.default), rawToFinal('polkadot', _polkadot.default), rawToFinal('westend', _westend.default)];
var _default = upgrades;
exports.default = _default;
