// @polkadot/types-known/upgrades/index.mjs
// Copyright 2017-2021 @polkadot/types-known authors & contributors
// SPDX-License-Identifier: Apache-2.0
import BN from 'bn.js';
import { assert, hexToU8a } from '@polkadot/util';
import kusama from "./kusama.mjs";
import polkadot from "./polkadot.mjs";
import westend from "./westend.mjs";
/** @internal */

function checkOrder(network, versions) {
  const ooo = versions.filter((curr, index) => {
    const prev = versions[index - 1];
    return index === 0 ? false : curr[0] <= prev[0] || curr[1] <= prev[1];
  });
  assert(!ooo.length, `${network}: Mismatched upgrade ordering: ${JSON.stringify(ooo)}`);
  return versions;
}
/** @internal */


function rawToFinal(network, {
  genesisHash,
  versions
}) {
  return {
    genesisHash: hexToU8a(genesisHash),
    network,
    versions: checkOrder(network, versions).map(([blockNumber, specVersion]) => ({
      blockNumber: new BN(blockNumber),
      specVersion: new BN(specVersion)
    }))
  };
} // Type overrides for specific spec types & versions as given in runtimeVersion


const upgrades = [rawToFinal('kusama', kusama), rawToFinal('polkadot', polkadot), rawToFinal('westend', westend)];
export default upgrades;
