// @polkadot/util-crypto/key/DeriveJunction.mjs
import _classPrivateFieldLooseBase from "@babel/runtime/helpers/esm/classPrivateFieldLooseBase";
import _classPrivateFieldLooseKey from "@babel/runtime/helpers/esm/classPrivateFieldLooseKey";
// Copyright 2017-2021 @polkadot/util-crypto authors & contributors
// SPDX-License-Identifier: Apache-2.0
import { bnToHex, compactAddLength, hexToU8a, isBigInt, isBn, isHex, isNumber, isString, stringToU8a } from '@polkadot/util';
import { blake2AsU8a } from "../blake2/asU8a.mjs";
const RE_NUMBER = /^\d+$/;
const JUNCTION_ID_LEN = 32;
const BN_OPTIONS = {
  bitLength: 256,
  isLe: true
};

var _chainCode = _classPrivateFieldLooseKey("chainCode");

var _isHard = _classPrivateFieldLooseKey("isHard");

export class DeriveJunction {
  constructor() {
    Object.defineProperty(this, _chainCode, {
      writable: true,
      value: new Uint8Array(32)
    });
    Object.defineProperty(this, _isHard, {
      writable: true,
      value: false
    });
  }

  static from(value) {
    const result = new DeriveJunction();
    const [code, isHard] = value.startsWith('/') ? [value.substr(1), true] : [value, false];
    result.soft(RE_NUMBER.test(code) ? parseInt(code, 10) : code);
    return isHard ? result.harden() : result;
  }

  get chainCode() {
    return _classPrivateFieldLooseBase(this, _chainCode)[_chainCode];
  }

  get isHard() {
    return _classPrivateFieldLooseBase(this, _isHard)[_isHard];
  }

  get isSoft() {
    return !_classPrivateFieldLooseBase(this, _isHard)[_isHard];
  }

  hard(value) {
    return this.soft(value).harden();
  }

  harden() {
    _classPrivateFieldLooseBase(this, _isHard)[_isHard] = true;
    return this;
  }

  soft(value) {
    if (isNumber(value) || isBn(value) || isBigInt(value)) {
      return this.soft(bnToHex(value, BN_OPTIONS));
    } else if (isString(value)) {
      return isHex(value) ? this.soft(hexToU8a(value)) : this.soft(compactAddLength(stringToU8a(value)));
    }

    if (value.length > JUNCTION_ID_LEN) {
      return this.soft(blake2AsU8a(value));
    }

    _classPrivateFieldLooseBase(this, _chainCode)[_chainCode].fill(0);

    _classPrivateFieldLooseBase(this, _chainCode)[_chainCode].set(value, 0);

    return this;
  }

  soften() {
    _classPrivateFieldLooseBase(this, _isHard)[_isHard] = false;
    return this;
  }

}
